operators_logic.go

```diff
@@ -0,0 +1,77 @@
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/corazawaf/coraza/v3/experimental/plugins/plugintypes"
+	"github.com/oschwald/geoip2-golang"
+)
+
+type geo struct {
+	db     geoIPReaderShim
+	dbtype string
+}
+
+func newGeolookupCreator(db *geoip2.Reader, databaseType string) func(options plugintypes.OperatorOptions) (plugintypes.Operator, error) {
+	return func(options plugintypes.OperatorOptions) (plugintypes.Operator, error) {
+		return newGeolookup(options, db, databaseType)
+	}
+}
+
+func newGeolookup(options plugintypes.OperatorOptions, db *geoip2.Reader, databaseType string) (plugintypes.Operator, error) {
+	return &geo{db: db, dbtype: databaseType}, nil
+}
+
+func (o *geo) ApplyVariablesCity(col collectionShim, ip net.IP) (bool, error) {
+	r, err := o.db.City(ip)
+	if err != nil {
+		return false, err
+	}
+
+	col.Set("country_code", []string{r.Country.IsoCode})
+	col.Set("country_name", []string{r.Country.Names["en"]})
+	col.Set("continent_code", []string{r.Continent.Code})
+	col.Set("country_continent", []string{r.Continent.Names["en"]})
+	col.Set("region", []string{""})
+	col.Set("city", []string{r.City.Names["en"]})
+	col.Set("postal_code", []string{r.Postal.Code})
+	col.Set("latitude", []string{strconv.FormatFloat(r.Location.Latitude, 'f', 10, 64)})
+	col.Set("longitude", []string{strconv.FormatFloat(r.Location.Longitude, 'f', 10, 64)})
+
+	return true, nil
+}
+
+func (o *geo) ApplyVariablesCountry(col collectionShim, ip net.IP) (bool, error) {
+	r, err := o.db.Country(ip)
+	if err != nil {
+		return false, err
+	}
+
+	col.Set("country_code", []string{r.Country.IsoCode})
+	col.Set("country_name", []string{r.Country.Names["en"]})
+	col.Set("continent_code", []string{r.Continent.Code})
+	col.Set("country_continent", []string{r.Continent.Names["en"]})
+
+	return true, nil
+}
+
+func (o *geo) executeEvaluationInternal(tx txShim, value string) (bool, error) {
+	col, err := tx.GetGeoCollection(tx)
+	if err != nil {
+		return false, err
+	}
+	ip := net.ParseIP(value)
```

Comment:
I wonder two things here for the sake of early return:
1. if the valye is empty, shall we continue?
2. if the IP is private, shall we continue?

Reply:
Good question, I would say for 1. We should not continue, for the 2. We should.

Reply:
CC @fzipi @jptosso

Reply:
I think for both cases we should not continue. Also, there might be more border cases that we might want to take into account. What about [`0.0.0.0`](https://pkg.go.dev/net/netip#IPv4Unspecified) or `255.255.255.0`?

Reply:
@fzipi I just added the skipping of specific ip ranges. I'm not sure if we should return `false` or `true` in that case. `false` means the filter failed, and the request gets denied, correct?